### PR TITLE
Librarian projections issue 8

### PR DESCRIPTION
**Pull request by PaulAgeyev**
Create projections for librarian:

-  AddBookView
-  AppendInventoryView
-  BookCompactHistory
-  BookFullHistory
-  DetailsView
-  ExtendedInventoryView
-  InventoryByBookView
-  InventoryView
-  LostBooksView
-  ReadersView
-  ReservationView
-  UserHistoryExtendedView
-  UserHistoryView

Add 'Year' - year of publication of the boor in values.proto.

### CODE REVIEW

File: model/src/main/proto/javaclasses/exlibris/q/projections.proto [hunk 1]
```diff
@@ -87,3 +89,252 @@ message HomePage {
...
+}
+
+// View with information about book’s loan.
+//
+// View consists of fields about inventory item,
+// users who borrowed the book and their taken on and due on dates.
+//
+message LoanDetails {
+
+    // An identifier of a book. Contains ISBN base62.
```
Comment (Dmytry-Dyachenko) [inline model/src/main/proto/javaclasses/exlibris/q/projections.proto]: The identifier of item contains not only ISBN with base62, it is ISBN and number that inputted by librarian too.
Use the comment from the line 193.

File: model/src/main/proto/javaclasses/exlibris/q/projections.proto [hunk 2]
```diff
@@ -87,3 +89,252 @@ message HomePage {
...
+// View with information about inventory item record.
+//
+// View consists of fields about inventory item, actions and action date.
+//
+message InventoryItemRecord {
+
+    // An identifier is represented by BookId and item number.
+    InventoryItemId inventory_item_id = 1;
+
+    // One of the actions is assigned to the intevntory item.
```
Comment (Dmytry-Dyachenko) [inline model/src/main/proto/javaclasses/exlibris/q/projections.proto]: intevntory -> inventory

File: model/src/main/proto/javaclasses/exlibris/q/projections.proto [hunk 3]
```diff
@@ -87,3 +89,252 @@ message HomePage {
...
+    oneof bookrecord {
```
Comment (Dmytry-Dyachenko) [inline model/src/main/proto/javaclasses/exlibris/q/projections.proto]: bookrecord -> book_record

File: model/src/main/proto/javaclasses/exlibris/q/projections.proto [hunk 4]
```diff
@@ -87,3 +89,252 @@ message HomePage {
...
+        bool lost = 6;
+    }
+
+    // The time when the action was did.
+    google.protobuf.Timestamp action_date = 7;
+}
+
+// Details about book.
+//
+// View consists of all about book.
```
Comment (Dmytry-Dyachenko) [inline model/src/main/proto/javaclasses/exlibris/q/projections.proto]: View consists of all about book -> View consists of all information about book

File: model/src/main/proto/javaclasses/exlibris/q/projections.proto [hunk 5]
```diff
@@ -87,3 +89,252 @@ message HomePage {
...
+// Url cover, book title, authors, isbn, year of publication,
```
Comment (Dmytry-Dyachenko) [inline model/src/main/proto/javaclasses/exlibris/q/projections.proto]: Url -> URL

File: model/src/main/proto/javaclasses/exlibris/q/projections.proto [hunk 6]
```diff
@@ -87,3 +89,252 @@ message HomePage {
...
+    // List of categories.
+    repeated Category categories = 6;
+
+    // Full book’s description.
+    BookDescription book_desctiption = 7;
+}
+
+// View contains a book loan in inventory.
+//
+// View consists of fields about inventory items,
```
Comment (Dmytry-Dyachenko) [inline model/src/main/proto/javaclasses/exlibris/q/projections.proto]: View consists of fields about inventory items, -> View consists of fields with information about inventory items,

File: model/src/main/proto/javaclasses/exlibris/q/projections.proto [hunk 7]
```diff
@@ -87,3 +89,252 @@ message HomePage {
...
+//
+message ExtendedInventoryView {
+
+    // List with information about book’s loan.
+    repeated LoanDetails loan_details = 1;
+}
+
+// Details about inventory.
+//
+// View consists of fields about book title, authors,
```
Comment (Dmytry-Dyachenko) [inline model/src/main/proto/javaclasses/exlibris/q/projections.proto]: +// View consists of fields about book title, authors, -> +// View consists of fields `with` book title, authors,

File: model/src/main/proto/javaclasses/exlibris/q/projections.proto [hunk 8]
```diff
@@ -87,3 +89,252 @@ message HomePage {
...
+// available books, amount overdue book.
```
Comment (Dmytry-Dyachenko) [inline model/src/main/proto/javaclasses/exlibris/q/projections.proto]: amount overdue book. -> amount of overdue books.

File: model/src/main/proto/javaclasses/exlibris/q/projections.proto [hunk 9]
```diff
@@ -87,3 +89,252 @@ message HomePage {
...
+    // Full author’s name.
+    repeated Author authors = 2;
+
+    // Amount of available books.
+    int32 availableItems = 3;
+
+    // Amount of books.
+    int32 itemsSize = 4;
+
+    // Amount overdue book.
```
Comment (Dmytry-Dyachenko) [inline model/src/main/proto/javaclasses/exlibris/q/projections.proto]: Amount overdue book. ->  Amount of overdue books.

File: model/src/main/proto/javaclasses/exlibris/q/projections.proto [hunk 10]
```diff
@@ -87,3 +89,252 @@ message HomePage {
...
+    // URL for book cover.
+    spine.net.Url url = 1;
+
+    // Information about book’s loan.
+    repeated LoanDetails loan_details = 3;
+}
+
+// View with information about inventory.
+//
+// View consists of fields about book title, authors,
```
Comment (Dmytry-Dyachenko) [inline model/src/main/proto/javaclasses/exlibris/q/projections.proto]: View consists of fields about book title, authors, ->  View consists of fields with book title, authors,

**Review by Dmytry-Dyachenko — CHANGES_REQUESTED** [review]
Please, edit all changes.

File: model/src/main/proto/javaclasses/exlibris/q/projections.proto [hunk 11]
```diff
@@ -87,3 +89,252 @@ message HomePage {
     repeated Category category = 2;
 }
 // *** This is projections for UI role librarian ***
+// View for adding book.
+//
+// View consists of fields for writing book details.
+//
+message AddBookView {
+
+    //List of book categories.
```
Comment (alexanderkarpetc) [inline model/src/main/proto/javaclasses/exlibris/q/projections.proto]: Add whitespace

File: model/src/main/proto/javaclasses/exlibris/q/projections.proto [hunk 12]
```diff
@@ -87,3 +89,252 @@ message HomePage {
...
+    repeated Category categories = 1;
+}
+
+// View for editing book.
```
Comment (alexanderkarpetc) [inline model/src/main/proto/javaclasses/exlibris/q/projections.proto]: editing -> updating

File: model/src/main/proto/javaclasses/exlibris/q/projections.proto [hunk 13]
```diff
@@ -87,3 +89,252 @@ message HomePage {
...
+    // Full author’s name.
+    repeated Author authors = 4;
+
+    // Year of publication of the book.
+    Year year = 5;
+
+    // Categories of book.
+    repeated Category categories = 6;
+
+    // Full book’s description.
```
Comment (alexanderkarpetc) [inline model/src/main/proto/javaclasses/exlibris/q/projections.proto]: book's -> book

File: model/src/main/proto/javaclasses/exlibris/q/projections.proto [hunk 14]
```diff
@@ -87,3 +89,252 @@ message HomePage {
...
+    BookDescription description = 7;
+}
+
+// View for appending inventory.
+//
+// The second step of wizard of adding book items.
+//
+message AppendInventoryView {
+
+    // An identifier is represented by BookId.
```
Comment (alexanderkarpetc) [inline model/src/main/proto/javaclasses/exlibris/q/projections.proto]: use backticks

File: model/src/main/proto/javaclasses/exlibris/q/projections.proto [hunk 15]
```diff
@@ -87,3 +89,252 @@ message HomePage {
...
+    InventoryId inventory_id = 2;
+}
+
+// View with compact information about book’s loans.
```
Comment (alexanderkarpetc) [inline model/src/main/proto/javaclasses/exlibris/q/projections.proto]: book's -> book

File: model/src/main/proto/javaclasses/exlibris/q/projections.proto [hunk 16]
```diff
@@ -87,3 +89,252 @@ message HomePage {
...
+    // URL for book cover.
+    spine.net.Url url = 1;
+
+    // List of user's loan details.
+    repeated LoanDetails loans = 2;
+}
+
+// View with information about book’s loan.
+//
+// View consists of fields about inventory item,
```
Comment (alexanderkarpetc) [inline model/src/main/proto/javaclasses/exlibris/q/projections.proto]: delete `about`

File: model/src/main/proto/javaclasses/exlibris/q/projections.proto [hunk 17]
```diff
@@ -87,3 +89,252 @@ message HomePage {
...
+    repeated InventoryItemRecord inventory_item_record = 2;
+}
+
+// View with information about inventory item record.
+//
+// View consists of fields about inventory item, actions and action date.
+//
+message InventoryItemRecord {
+
+    // An identifier is represented by BookId and item number.
```
Comment (alexanderkarpetc) [inline model/src/main/proto/javaclasses/exlibris/q/projections.proto]: backticks, please

File: model/src/main/proto/javaclasses/exlibris/q/projections.proto [hunk 18]
```diff
@@ -87,3 +89,252 @@ message HomePage {
...
+message InventoryDetails {
+
+    // Book’s title.
+    BookTitle book_title = 1;
+
+    // Full author’s name.
+    repeated Author authors = 2;
+
+    // Amount of available books.
+    int32 availableItems = 3;
```
Comment (alexanderkarpetc) [inline model/src/main/proto/javaclasses/exlibris/q/projections.proto]: available_items

File: model/src/main/proto/javaclasses/exlibris/q/projections.proto [hunk 19]
```diff
@@ -87,3 +89,252 @@ message HomePage {
...
+
+    // Amount of books.
+    int32 itemsSize = 4;
```
Comment (alexanderkarpetc) [inline model/src/main/proto/javaclasses/exlibris/q/projections.proto]: same

File: model/src/main/proto/javaclasses/exlibris/q/projections.proto [hunk 20]
```diff
@@ -87,3 +89,252 @@ message HomePage {
...
+    spine.net.Url url = 1;
+
+    // Information about book’s loan.
+    repeated LoanDetails loan_details = 3;
+}
+
+// View with information about inventory.
+//
+// View consists of fields with book title, authors,
+// available books, amount overdue book.
```
Comment (alexanderkarpetc) [inline model/src/main/proto/javaclasses/exlibris/q/projections.proto]: amount of

File: model/src/main/proto/javaclasses/exlibris/q/projections.proto [hunk 21]
```diff
@@ -87,3 +89,252 @@ message HomePage {
...
+//
+message InventoryView {
+
+    // Details about inventory.
```
Comment (alexanderkarpetc) [inline model/src/main/proto/javaclasses/exlibris/q/projections.proto]: delete about

**Review by alexanderkarpetc — CHANGES_REQUESTED** [review]
Please fix mistakes

File: model/src/main/proto/javaclasses/exlibris/q/projections.proto [hunk 22]
```diff
@@ -87,3 +89,252 @@ message HomePage {
...
+//
+message AddBookView {
+
+    //List of book categories.
+    repeated Category categories = 1;
+}
+
+// View for editing book.
+//
+// View consists of fields for edit book details.
```
Comment (Dmytry-Dyachenko) [inline model/src/main/proto/javaclasses/exlibris/q/projections.proto]: edit - > updating

File: model/src/main/proto/javaclasses/exlibris/q/projections.proto [hunk 23]
```diff
@@ -87,3 +89,252 @@ message HomePage {
...
+    // Book’s title.
+    BookTitle title = 3;
+
+    // Full author’s name.
+    repeated Author authors = 4;
+
+    // Year of publication of the book.
+    Year year = 5;
+
+    // Categories of book.
```
Comment (Dmytry-Dyachenko) [inline model/src/main/proto/javaclasses/exlibris/q/projections.proto]: Book categories.

File: model/src/main/proto/javaclasses/exlibris/q/projections.proto [hunk 24]
```diff
@@ -87,3 +89,252 @@ message HomePage {
...
+//
+// Librarian sees all users that borrow book,
+// reserve, lost, return, or extend loan period with date of every record.
+//
+message BookFullHistory {
+
+    // URL for book cover.
+    spine.net.Url url = 1;
+
+    //
```
Comment (Dmytry-Dyachenko) [inline model/src/main/proto/javaclasses/exlibris/q/projections.proto]: Docs are absent.

File: model/src/main/proto/javaclasses/exlibris/q/projections.proto (hunk 17 above)
Comment (Dmytry-Dyachenko) [inline model/src/main/proto/javaclasses/exlibris/q/projections.proto]: add backticks

**Review by Dmytry-Dyachenko — CHANGES_REQUESTED** [review]
Fix mistakes please.,

**Review by alexanderkarpetc — APPROVED** [review]
LGTM

**Review by Dmytry-Dyachenko — APPROVED** [review]
LGTM

**Pull request merged**